### PR TITLE
Fix on-demand initialization race conditions [marshal.c].

### PR DESCRIPTION
**Pull request by jaykrell**
Extracted from https://github.com/mono/mono/pull/18150 which reviewer said was too big.

### CODE REVIEW

**Comment by jaykrell** [comment]
@monojenkins build failed

File: mono/metadata/marshal.c [hunk 1]
```diff
@@ -4257,11 +4256,17 @@ mono_marshal_get_struct_to_ptr (MonoClass *klass)
 	if (marshal_info->str_to_ptr)
 		return marshal_info->str_to_ptr;
 
-	if (!stoptr) {
+	if ((res = marshal_info->str_to_ptr))
+		return res;
```
Comment (lambdageek) [inline mono/metadata/marshal.c]: Duplicating logic from the preceeding `if`. Guessing you meant to delete one of them.

Comment (jaykrell) [reply mono/metadata/marshal.c]: Yeah it was a "bad" merge/rebase, thank you.

Comment (jaykrell) [reply mono/metadata/marshal.c]: (or checkout -p of old revision, similar)

File: mono/metadata/marshal.c [hunk 2]
```diff
@@ -4257,11 +4256,17 @@ mono_marshal_get_struct_to_ptr (MonoClass *klass)
 	if (marshal_info->str_to_ptr)
 		return marshal_info->str_to_ptr;
 
-	if (!stoptr) {
+	if ((res = marshal_info->str_to_ptr))
```
[the remark this answers is not in the document]
Comment (jaykrell) [reply mono/metadata/marshal.c]: It was a bad merge, @lambdageek noticed too, should already appear fixed.

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed

**Pull request merged**